client/protoclient.go

```diff
@@ -0,0 +1,523 @@
+	}
+}
+
+// Receives one json string from the auto documentation, decode it and request
+// to the server the protobuf descriptors. If the the  descriptors route are
+// not set, this function identify the route responsible for providing the
+// protobuf descriptors.
+func (pc *ProtoClient) getDescriptors(data string) error {
+	d := []byte(data)
+	var jsonmap interface{}
+	if err := json.Unmarshal(d, &jsonmap); err != nil {
+		return err
+	}
+	m := jsonmap.(map[string]interface{})
+	keysSet := make(map[string]bool)
+	getKeys(m, keysSet)
+
+	// load predefined protos
+	for _, commands := range pc.info.Commands {
+		if commands.input != "" {
+			keysSet[commands.input] = true
+		}
+		if commands.output != "" {
+			keysSet[commands.output] = true
+		}
+	}
+
+	// build commands reference
+	handlers := m["handlers"].(map[string]interface{})
+	for k, v := range handlers {
+		cmdInfo := v.(map[string]interface{})
+		in, out, err := getOutputInputNames(cmdInfo)
+		if err != nil {
+			return err
+		}
+
+		var command Command
+		command.input = in
+		command.output = out
+
+		pc.info.Commands[k] = &command
+		if pc.descriptorsRoute == "" && in == "protos.ProtoNames" && out == "protos.ProtoDescriptors" {
+			pc.descriptorsRoute = k
+		}
+	}
+
+	remotes := m["remotes"].(map[string]interface{})
+	for k, v := range remotes {
+		cmdInfo := v.(map[string]interface{})
+		in, out, err := getOutputInputNames(cmdInfo)
+		if err != nil {
+			return err
+		}
+
+		var command Command
+		command.input = in
+		command.output = out
+
+		pc.info.Commands[k] = &command
+	}
+
+	// get all proto types
+	descriptorArray := make([]*protobuf.FileDescriptorProto, 0)
+
+	var names []string
+
+	for key := range keysSet {
+		names = append(names, key)
+	}
+
+	protname := &protos.ProtoNames{
+		Name: names,
+	}
+
+	encodedNames, err := proto.Marshal(protname)
+	if err != nil {
+		return err
+	}
+	_, err = pc.SendRequest(pc.descriptorsRoute, encodedNames)
+	if err != nil {
+		return err
+	}
+
+	response := <-pc.Client.IncomingMsgChan
+
+	descriptors := &protos.ProtoDescriptors{}
+
+	if err := proto.Unmarshal(response.Data, descriptors); err != nil {
+		return err
+	}
+
+	for i := range descriptors.Desc {
+		fileDescriptorProto, err := unpackDescriptor(descriptors.Desc[i])
+		if err != nil {
+			return err
+		}
+
+		descriptorArray = append(descriptorArray, fileDescriptorProto)
+		pc.descriptorsNames[names[i]] = true
+	}
+
+	err = pc.buildProtosFromDescriptor(descriptorArray)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Return the basic structure for the ProtoClient struct.
+func newProto(docslogLevel logrus.Level, requestTimeout ...time.Duration) *ProtoClient {
+	return &ProtoClient{
+		Client:           *New(docslogLevel, requestTimeout...),
+		descriptorsNames: make(map[string]bool),
+		info: ProtoBufferInfo{
+			Commands: make(map[string]*Command),
+		},
+		docsRoute:        "",
+		descriptorsRoute: "",
+		IncomingMsgChan:  make(chan *message.Message, 3),
```

Comment:
Any special reason for the value 3?

Reply:
The main reason is to create a small buffer. I will change it to 10, to be equal to the client/client.go